libraries/Parsers/Microsoft.Bot.Builder.Parsers.LU/Parser/Visitor.cs

```diff
@@ -0,0 +1,207 @@
+                    {
+                        errorMessages.Add($"Invalid utterance line, entity name {entityName} cannot contain any of the following characters: [<, >, *, %, &, :, \\, $]");
+                        continue;
+                    }
+
+                    if (auxEntity.EntityValue == null)
+                    {
+                        // we have a pattern.any entity
+                        var patternStr = !string.IsNullOrEmpty(auxEntity.Role) ? $"{{{auxEntity.EntityName}:{auxEntity.Role}}}" : $"{{{auxEntity.EntityName}}}";
+                        srcUtterance += patternStr;
+                        entities.Add(
+                            new EntityElement
+                            {
+                                Type = TypeEnum.PatternAnyEntities,
+                                Entity = auxEntity.EntityName.Trim(),
+                                Role = auxEntity.Role.Trim()
+                            });
+                    }
+                    else
+                    {
+                        // we have a new entity
+                        var newEntity = new EntityElement
+                        {
+                            Type = TypeEnum.Entities,
+                            Entity = auxEntity.EntityName.Trim(),
+                            Role = auxEntity.Role.Trim(),
+                            StartPos = srcUtterance.Length,
+                            EndPos = null
+                        };
+                        if (auxEntity.EntityValue == null)
+                        {
+                            errorMessages.Add($"Composite entity {auxEntity.Parent.EntityName} includes pattern.any entity {auxEntity.EntityName}.\r\n\tComposites cannot include pattern.any entity as a child.");
+                        }
+                        else
+                        {
+                            srcUtterance = RecursivelyResolveTokenizedUtterance(auxEntity.EntityValue, entities, errorMessages, srcUtterance).TrimStart();
+                            newEntity.EndPos = srcUtterance.Length - 1;
+                            entities.Add(newEntity);
+                        }
+                    }
+                }
+                else
+                {
+                    char charItem = (char)item;
+                    srcUtterance += charItem;
+                }
+            }
+
+            return srcUtterance;
+        }
+
+        private static List<object> TokenizeUtterance(string expression)
+        {
+            var splitString = new List<object>();
+            var currentList = splitString;
+            AuxEntity currentEntity = null;
+            var entityNameCapture = false;
+            var entityValueCapture = false;
+            var entityRoleCapture = false;
+
+            foreach (char character in expression)
+            {
+                switch (character)
+                {
+                    case '{':
```

Comment:
This is scary, and probably bug prone. Parsing like this never ends well. Is there another option? 

If nothing else, this code will struggle and have incorrect results with 16-bit Unicode codepoints that don't fit in "char". This should be StringInfo class, and doing parsing that way. Unicode is our friend. :)

Reply:
Solved